plugins/salesforce/src/org/pentaho/di/trans/steps/salesforce/SalesforceConnection.java

```diff
@@ -934,43 +925,45 @@ public static MessageElement createMessageElement( String name, Object value, bo
     return me;
   }
 
-  private static MessageElement createForeignKeyElement( String type, String lookupField, String extIdName,
+  private static XmlObject createForeignKeyElement( String type, String lookupField, String extIdName,
     Object extIdValue ) throws Exception {
 
     // Foreign key relationship to the object
-    MessageElement me = fromTemplateElement( lookupField, null, false );
-    me.addChild( new MessageElement( new QName( "type" ), type ) );
-    me.addChild( new MessageElement( new QName( extIdName ), extIdValue ) );
+    XmlObject me = fromTemplateElement( lookupField, null, false );
+    me.addField( "type", type );
+    me.addField( extIdName, extIdValue );
 
     return me;
   }
 
-  private static MessageElement TEMPLATE_MESSAGE_ELEMENT = new MessageElement( "", "temp" );
-
-  // The Template org.w3c.dom.Element instance
-  private static Element TEMPLATE_XML_ELEMENT;
-
-  static {
-    try {
-      // Create and cache this org.w3c.dom.Element instance for once here.
-      TEMPLATE_XML_ELEMENT = TEMPLATE_MESSAGE_ELEMENT.getAsDOM();
-    } catch ( Exception e ) {
-      throw new RuntimeException( e );
-    }
-    TEMPLATE_XML_ELEMENT.removeAttribute( "xsi:type" );
-    TEMPLATE_XML_ELEMENT.removeAttribute( "xmlns:ns1" );
-    TEMPLATE_XML_ELEMENT.removeAttribute( "xmlns:xsd" );
-    TEMPLATE_XML_ELEMENT.removeAttribute( "xmlns:xsi" );
-  }
-
-  public static MessageElement fromTemplateElement( String name, Object value, boolean setValue ) throws SOAPException {
+  public static XmlObject fromTemplateElement( String name, Object value, boolean setValue ) throws SOAPException {
     // Use the TEMPLATE org.w3c.dom.Element to create new Message Elements
-    MessageElement me = new MessageElement( TEMPLATE_XML_ELEMENT );
+    XmlObject me = new XmlObject();
     if ( setValue ) {
-      me.setObjectValue( value );
+      me.setValue( value );
     }
-    me.setName( name );
+    me.setName( new QName( name ) );
     return me;
   }
 
+  public static XmlObject[] getChildren( SObject object ) {
+    List<String> reservedFieldNames = Arrays.asList( "type", "fieldsToNull" );
+    if ( object == null ) {
+      return null;
+    }
+    List<XmlObject> children = new ArrayList<>();
+    Iterator<XmlObject> iterator = object.getChildren();
+    while ( iterator.hasNext() ) {
+      XmlObject child = iterator.next();
+      if ( child.getName().getNamespaceURI().equals( "urn:sobject.partner.soap.sforce.com" )
```

Comment:
I would like to see this hard-coded string ( "urn:sobject.partner.soap.sforce.com" ) moved to a property read from the kettle.properties so that when (not if) sf.com decides to change that, we don't have to rebuild for existing customers.